src/cargo/core/dependency.rs

```diff
@@ -171,8 +175,8 @@ impl Dependency {
         let name = name.into();
         let arg = Some((inside, config));
         let (specified_req, version_req) = match version {
-            Some(v) => (true, parse_req_with_deprecated(name, v, arg)?),
-            None => (false, VersionReq::any()),
+            Some(v) => (true, parse_req_with_deprecated(name, v, arg)?.into()),
+            None => (false, OptVersionReq::Any),
```

Comment:
"Any" is not considered a VersionReq, because there is no semver syntax that literally matches any version. `*` is not that, because of how https://github.com/semver/semver/blob/efcff2c838c9945f79bfd21c1df0073271bcc29c/ranges.md suggests handling pre-release versions:

> - SemVer Versions containing PRERELEASE identifiers MUST NOT be included by a SemVer Comparator Set unless they are included by all of the Comparators in the Comparator Set, and one or more of the following conditions are met:
>
>   - The implementation has provided the user with an option to explicitly include PRERELEASE versions, and the user has set this option. This SHOULD NOT be the default behavior when resolving software dependencies.
>
>   - One or more of the Comparators in the Comparator Set have the same MAJOR, MINOR, and PATCH version as the SemVer Version string being considered, and has a PRERELEASE version, and would normally include the SemVer Version string by virtue of precedence comparison.
>
>       For example, the Range >=1.0.0-alpha would include the Version 1.0.0-beta but not the Version 1.0.1-beta.

In previous versions of the semver crate, `VersionReq::any()` would literally mean any, which is different from `*`, and it would match *any* version. But upon roundtrip through a string, it would decay to `*`, because again there is no semver syntax that means "any".

Instead I'm suggesting that Cargo use something effectively like `Option<VersionReq>` with an absent VersionReq meaning "any", and the equivalent of `.map_or(true, |req| req.matches(version))` for matching.